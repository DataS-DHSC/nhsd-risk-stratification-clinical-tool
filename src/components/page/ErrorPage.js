import Head from 'next/head';
import PropTypes from 'prop-types';
import cookie from 'react-cookies';
import { ROUTE_IP_ENDPOINT, ROUTE_SUPPORT_FORM } from '../../constants/routes';
import { SITE_BRANDING_TITLE } from '../../constants/site-branding';

const errors = [
  {
    code: 400,
    title: 'Something went wrong',
  },
  {
    code: 404,
    title: 'Page not found',
    message:
      'If you entered a web address, make sure you entered it correctly.\n\nIf you pasted the web address, make sure you copied the entire address. ',
  },
  {
    code: 503,
    title: 'Service unavailable',
    message:
      'Sorry, but this service is currently unavailable whilst a new version is being released. It will be available again soon. Please try again later.',
  },
];

const ErrorPage = ({ statusCode, startAgainLink }) => {
  let error = errors.find((e) => e.code === statusCode);
  if (!error) error = errors.find((e) => e.code === 400);

  const correlationId = cookie.load('correlationId') || 'ead1c4fe';

  return (
    <>
      <Head>
        <title>{SITE_BRANDING_TITLE} - Error</title>
        <meta name="description" content={`${SITE_BRANDING_TITLE} error`} />
        <meta name="viewport" content="initial-scale=1.0, width=device-width" />
      </Head>
      <div className="nhsuk-grid-row">
        <div className="nhsuk-grid-column-full">
          <h1 id="error-heading">{error.title}</h1>
        </div>
        <div className="nhsuk-grid-column-two-thirds">
          {statusCode === 400 || statusCode === 403 ? (
            <p id="temporary-problem-message">
              Sorry, an error has occurred with the service. This is usually a
              temporary problem, but if it continues you should let our Helpdesk
              know.
              <br />
              <br />
              If you need to contact the Helpdesk:
              <br />
              <br />
              You may first need to{' '}
              <a
                href={ROUTE_IP_ENDPOINT}
                rel="noreferrer nofollow noopener"
                target="_blank"
              >
                find your IP address (opens in new window)
              </a>
              {correlationId &&
                ` and enter that as well as your Correlation ID: ${correlationId} `}
              in ‘Your message for us’ when you{' '}
              <a
                rel="noreferrer nofollow noopener"
                target="_blank"
                href={encodeURI(`${ROUTE_SUPPORT_FORM}`)}
                id="helpdesk-link"
              >
                complete the helpdesk contact form (opens in new window)
              </a>
            </p>
          ) : (
            <p style={{ whiteSpace: 'pre-wrap' }}>{error.message}</p>
          )}

          {startAgainLink && (
            <a
              href={startAgainLink}
              className="nhsuk-u-margin-top-5 nhsuk-u-font-size-19"
              id="new-assessment-link"
            >
              Start a new COVID-19 risk assessment
            </a>
          )}
        </div>
      </div>
    </>
  );
};

ErrorPage.propTypes = {
  statusCode: PropTypes.number.isRequired,
  startAgainLink: PropTypes.string,
};

ErrorPage.defaultProps = {
  startAgainLink: '',
};

export default ErrorPage;
