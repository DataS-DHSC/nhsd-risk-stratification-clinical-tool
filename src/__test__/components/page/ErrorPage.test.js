import React from 'react';
import { render } from '@testing-library/react';
import ErrorPage from '../../../components/page/ErrorPage';

const setup = (statusCode) => {
  const props = {
    statusCode,
  };

  return render(<ErrorPage {...props} />);
};

describe('ErrorPage', () => {
  it('should match snapshot when 400', () => {
    const container = setup(400);

    expect(container).toMatchSnapshot();
  });

  it('should match snapshot when 403', () => {
    const container = setup(403);

    expect(container.asFragment()).toMatchSnapshot();
  });

  it('should match snapshot when 404', () => {
    const container = setup(404);

    expect(container.asFragment()).toMatchSnapshot();
  });

  it('should match snapshot when 503', () => {
    const container = setup(503);

    expect(container.asFragment()).toMatchSnapshot();
  });
});
