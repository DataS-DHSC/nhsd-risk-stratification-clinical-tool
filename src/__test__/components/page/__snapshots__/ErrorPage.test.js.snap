// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorPage should match snapshot when 400 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="nhsuk-grid-row"
      >
        <div
          class="nhsuk-grid-column-full"
        >
          <h1
            id="error-heading"
          >
            Something went wrong
          </h1>
        </div>
        <div
          class="nhsuk-grid-column-two-thirds"
        >
          <p
            id="temporary-problem-message"
          >
            Sorry, an error has occurred with the service. This is usually a temporary problem, but if it continues you should let our Helpdesk know.
            <br />
            <br />
            If you need to contact the Helpdesk:
            <br />
            <br />
            You may first need to
             
            <a
              href="/ras/ip"
              rel="noreferrer nofollow noopener"
              target="_blank"
            >
              find your IP address (opens in new window)
            </a>
             and enter that as well as your Correlation ID: ead1c4fe 
            in ‘Your message for us’ when you
             
            <a
              href="https://www.nhs.uk/contact-us/covid-risk"
              id="helpdesk-link"
              rel="noreferrer nofollow noopener"
              target="_blank"
            >
              complete the helpdesk contact form (opens in new window)
            </a>
          </p>
          
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="nhsuk-grid-row"
    >
      <div
        class="nhsuk-grid-column-full"
      >
        <h1
          id="error-heading"
        >
          Something went wrong
        </h1>
      </div>
      <div
        class="nhsuk-grid-column-two-thirds"
      >
        <p
          id="temporary-problem-message"
        >
          Sorry, an error has occurred with the service. This is usually a temporary problem, but if it continues you should let our Helpdesk know.
          <br />
          <br />
          If you need to contact the Helpdesk:
          <br />
          <br />
          You may first need to
           
          <a
            href="/ras/ip"
            rel="noreferrer nofollow noopener"
            target="_blank"
          >
            find your IP address (opens in new window)
          </a>
           and enter that as well as your Correlation ID: ead1c4fe 
          in ‘Your message for us’ when you
           
          <a
            href="https://www.nhs.uk/contact-us/covid-risk"
            id="helpdesk-link"
            rel="noreferrer nofollow noopener"
            target="_blank"
          >
            complete the helpdesk contact form (opens in new window)
          </a>
        </p>
        
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`ErrorPage should match snapshot when 403 1`] = `
<DocumentFragment>
  <div
    class="nhsuk-grid-row"
  >
    <div
      class="nhsuk-grid-column-full"
    >
      <h1
        id="error-heading"
      >
        Something went wrong
      </h1>
    </div>
    <div
      class="nhsuk-grid-column-two-thirds"
    >
      <p
        id="temporary-problem-message"
      >
        Sorry, an error has occurred with the service. This is usually a temporary problem, but if it continues you should let our Helpdesk know.
        <br />
        <br />
        If you need to contact the Helpdesk:
        <br />
        <br />
        You may first need to 
        <a
          href="/ras/ip"
          rel="noreferrer nofollow noopener"
          target="_blank"
        >
          find your IP address (opens in new window)
        </a>
         and enter that as well as your Correlation ID: ead1c4fe in ‘Your message for us’ when you 
        <a
          href="https://www.nhs.uk/contact-us/covid-risk"
          id="helpdesk-link"
          rel="noreferrer nofollow noopener"
          target="_blank"
        >
          complete the helpdesk contact form (opens in new window)
        </a>
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ErrorPage should match snapshot when 404 1`] = `
<DocumentFragment>
  <div
    class="nhsuk-grid-row"
  >
    <div
      class="nhsuk-grid-column-full"
    >
      <h1
        id="error-heading"
      >
        Page not found
      </h1>
    </div>
    <div
      class="nhsuk-grid-column-two-thirds"
    >
      <p
        style="white-space: pre-wrap;"
      >
        If you entered a web address, make sure you entered it correctly.

If you pasted the web address, make sure you copied the entire address. 
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ErrorPage should match snapshot when 503 1`] = `
<DocumentFragment>
  <div
    class="nhsuk-grid-row"
  >
    <div
      class="nhsuk-grid-column-full"
    >
      <h1
        id="error-heading"
      >
        Service unavailable
      </h1>
    </div>
    <div
      class="nhsuk-grid-column-two-thirds"
    >
      <p
        style="white-space: pre-wrap;"
      >
        Sorry, but this service is currently unavailable whilst a new version is being released. It will be available again soon. Please try again later.
      </p>
    </div>
  </div>
</DocumentFragment>
`;
